cpp/src/arrow/acero/plan_test.cc

```diff
@@ -767,6 +767,24 @@ TEST(ExecPlanExecution, DeclarationToReader) {
                                   reader->Next());
 }
 
+TEST(ExecPlanExecution, DeclarationToReaderWithEarlyClose) {
+  auto random_data = MakeRandomBatches(schema({field("a", int8())}), /*num_batches=*/100);
+  auto plan = Declaration::Sequence(
+      {{"source", SourceNodeOptions(random_data.schema, random_data.gen(false, false))}});
+  ASSERT_OK_AND_ASSIGN(std::unique_ptr<RecordBatchReader> reader,
+                       DeclarationToReader(plan, /*use_threads=*/false));
+
+  // read only a few batches
+  for (size_t i = 0; i < 10; i++) {
+    ASSERT_OK(reader->Next());
+  }
+  // then close
+  EXPECT_RAISES_WITH_MESSAGE_THAT(Cancelled, HasSubstr("Plan was cancelled early"),
+                                  reader->Close());
```

Comment:
Reinstating this invocation requires a guarantee that `ExecPlan::StopProducing()` is idempotent, otherwise if `PlanReader::Close()` occurs then the invocation here which must occur as well may misbehave. I don't see this in guarantee in the documentation. @westonpace, WDYT?

Reply:
If the expectation of "close" here is to "cancel the query gracefully", then I think the reasonable behaviors should be returning status OK if the cancellation is successful. I don't think raises an exception "Plan is cancelled" is the correct behavior then the user actually wanted to "cancel the plan".

Reply:
The currently tested behavior exists pre-PR, so be aware that you're asking to change it. I don't know what the expectation in this case is because there is no documentation for `RecordBatchReader::Close()` and `PlanReader::Close()` that addresses this. There are other C++ reading APIs out there where a close operation is expected to report errors, possibly via a separate call ([like `std::ifstream::close()`](https://en.cppreference.com/w/cpp/io/basic_ifstream/close)), so I suspect people's expectations about this may vary. I'm fine with whatever behavior is in consensus, and in any case documentation of it should be added.

Reply:
I actually included the fix you requested, so please take a look. In any case, I'm fine with whatever is decided.